### PR TITLE
Completed quests widget

### LINKED ISSUE #1025: [Quests] Add completed quests to MyMeta profile
### What would you like to be added?

MyMeta profile should have a section which displays a list of all accepted quests.
A particular quest list item should include:
- title that links to the quest UI
- link to the proof of completion

### Todo

- [ ] make whole quest cards clickable (or just their headers)
- [x] figure out populating the `quest_completions` results with quest data
- [x] link the quests from profile section to their view
- [x] make the section pretty
- [x] make the `display all` modal

### Ideas

- xp rewarded

### Why is this needed?

Questing MVP https://miro.com/app/board/o9J_ldTAEnQ=/

### PR DESCRIPTION
**Pull request by vidvidvid**
## Overview

**What features/fixes does this PR include?**

Reusable widget of completed quests.
Displayed in user profile.

**Please provide the GitHub issue number**

Closes #1025 

## Follow up Improvement Ideas

- [ ] Add xp/seed gained
- [ ] Implement it in user dashboard

### Preview
Kapture 2022-01-30 at 18 55 34

[Figma](https://www.figma.com/proto/Uavi9O6xhYqGcK76fVP7bR/Profile-page?page-id=86%3A107&node-id=1322%3A2130&viewport=333%2C48%2C0.5&scaling=min-zoom&starting-point-node-id=1322%3A2130&show-proto-sidebar=1)

### CODE REVIEW

**Comment by vidvidvid** [comment]
something got messed up here, i need to fix this branch. sad life :'(

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:60]: Note that I switched from strings to numbers.

```suggestion
  const limitOptions = [10, 20, 50].map((n) => ({ label: n, value: n }))
```

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:111]:
```suggestion
        py={6}
```

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:116]:
```suggestion
        zIndex={1}
```

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:126]:
```suggestion
```

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:132]:
```suggestion
            const [v] = values.slice(-1);
```

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:135]:
```suggestion
              setQueryVariable('limit', v.value);
```

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:143]:
```suggestion
```

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:149]:
```suggestion
            const [o] = values.slice(-1);
```

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:166]:
```suggestion
            const [s] = values.slice(-1);
```

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:187]: After you've done the camel cased custom names mentioned at the beginning of this review.

```suggestion
                setQueryVariable('guildId', g.value);
```

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestFilter.tsx:202]: Does this work for setting the variable name and type? If so, I would use this format on all the filters.

```suggestion
            onChange={(values: Array<ValueType>) => {
```

File: packages/web/components/Quest/QuestForm.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestForm.tsx:25]:
```suggestion
import { RoleOption } from 'utils/roleHelpers';
```

File: packages/web/components/Quest/QuestForm.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestForm.tsx:152]:
```suggestion
```

File: packages/web/components/Quest/QuestForm.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestForm.tsx:350]:
```suggestion
                  {...{ roleChoices }}
```

File: packages/web/components/Quest/QuestTile.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestTile.tsx:84]:
```suggestion
            skills={quest.quest_skills.map(({ skill }) => skill) as Skill[]}
```

File: packages/web/components/Quest/QuestTile.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestTile.tsx:94]:
```suggestion
              roles={quest.quest_roles.map(({ PlayerRole: r }) => r) as PlayerRole[]}
```

File: packages/web/components/Quest/QuestForm.tsx
Comment (dysbulic) [inline packages/web/components/Quest/QuestForm.tsx:346]:
```suggestion
              {...{ control }}
              name="roles"
```

File: packages/web/components/Roles.tsx
Comment (dysbulic) [inline packages/web/components/Roles.tsx:25]: Does this work?

```suggestion
    onChange={(values: Array<RoleOption>) => setRoles(values)}
```

File: packages/web/components/Roles.tsx
Comment (dysbulic) [inline packages/web/components/Roles.tsx:46]:
```suggestion
    {roles.map(({ role, label, description }) => (
      <WrapItem key={role}>
        <Tooltip label={description}>
          <MetaTag size="md" fontWeight="normal" backgroundColor="purpleTag">
            {label}
```

File: packages/web/components/Setup/SetupRoles.tsx
Comment (dysbulic) [inline packages/web/components/Setup/SetupRoles.tsx:43]:
```suggestion
    roleChoices.filter(({ role, basic }) => !playerRoles.includes(role) && basic);
```

File: packages/web/graphql/getQuests.ts
Comment (dysbulic) [inline packages/web/graphql/getQuests.ts:37]:
```suggestion
    $questRoles: [String!]
```

File: packages/web/graphql/getQuests.ts
Comment (dysbulic) [inline packages/web/graphql/getQuests.ts:46]:
```suggestion
        quest_roles: { role: { _in: $questRoles } }
```

File: packages/web/graphql/getQuests.ts
Comment (dysbulic) [inline packages/web/graphql/getQuests.ts:117]:
```suggestion
```

File: packages/web/graphql/getQuests.ts
Comment (dysbulic) [inline packages/web/graphql/getQuests.ts:57]:
```suggestion
/* GraphQL */ `
```

File: packages/web/pages/profile/setup/roles.tsx
Comment (dysbulic) [inline packages/web/pages/profile/setup/roles.tsx:14]:
```suggestion
      roleChoices: roleChoices.filter(({ basic }) => basic),
```

File: packages/web/pages/quest/[id]/edit.tsx
Comment (dysbulic) [inline packages/web/pages/quest/[id]/edit.tsx:61]:
```suggestion
    const rolesObjects = data.roles.map(({ value }) => ({
      quest_id: quest.id,
      role: value,
```

File: packages/web/pages/quest/create.tsx
Comment (dysbulic) [inline packages/web/pages/quest/create.tsx:45]:
```suggestion
      skillIds: skills.map(({ id }) => id),
```

File: packages/web/pages/quest/create.tsx
Comment (dysbulic) [inline packages/web/pages/quest/create.tsx:46]:
```suggestion
      roleIds: roles.map(({ value }) => value),
```

File: packages/web/pages/quests.tsx
Comment (dysbulic) [inline packages/web/pages/quests.tsx:101]:
```suggestion
            {...{ roleChoices }}
```

File: packages/web/pages/quest/[id]/edit.tsx
Comment (dysbulic) [inline packages/web/pages/quest/[id]/edit.tsx:106]:
```suggestion
        {...{ roleChoices }}
```

**Review by dysbulic — CHANGES_REQUESTED** [review]
It isn't displaying correctly in the [test instance](//frontend-pr-1085-mjhnbmqqna-uk.a.run.app/), but the more recent versions haven't deployed because there are merge conflicts.

I didn't find any serious issues, just stylistic stuff.

I'll approve it once you fix the merge changes, & I can see it working.

Screenshot from 2022-02-02 13-46-03

File: packages/web/components/Quest/QuestFilter.tsx
Comment (vidvidvid) [reply packages/web/components/Quest/QuestFilter.tsx]: i'll use your syntax, but keep the strings, ok?

File: packages/web/components/Roles.tsx
Comment (vidvidvid) [reply packages/web/components/Roles.tsx]: getting
```
Type '(values: Array<RoleOption>) => void' is not assignable to type '(value: ValueType<{ label: string; value: string; }>, action: ActionMeta<{ label: string; value: string; }>) => void'.
  Types of parameters 'values' and 'value' are incompatible.
    Type 'ValueType<{ label: string; value: string; }>' is not assignable to type 'RoleOption[]'.
      Type 'undefined' is not assignable to type 'RoleOption[]'.ts(2322)
Select.d.ts(169, 3): The expected type comes from property 'onChange' which is declared here on type 'IntrinsicAttributes & Props<{ label: string; value: string; }> & { children?: ReactNode; }'
```
when i try to make this change

File: packages/web/components/Profile/ProfileSection.tsx
Comment (vidvidvid) [reply packages/web/components/Profile/ProfileSection.tsx:76]: I'm getting the following complaing if i remove boxType from conditions:
Screenshot 2022-02-02 at 21 24 46

File: packages/web/components/Quest/QuestDetails.tsx
Comment (vidvidvid) [reply packages/web/components/Quest/QuestDetails.tsx]: It could be, but I don't know how to do it, could you help me out? i'm first time working with hasura.

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [reply packages/web/components/Quest/QuestFilter.tsx]: One place you're using the values is in a string interpolation which will coerce it to a string. In the other you're converting the string to a number.

Why do you want to keep it as a string?

File: packages/web/components/Profile/ProfileSection.tsx
Comment (dysbulic) [reply packages/web/components/Profile/ProfileSection.tsx:76]: You'll have to loosen the type restrictions on `isBoxDataEditable`:

```typescript
const isBoxDataEditable = (boxType?: Maybe<BoxType>) =>
  !!boxType && [
    BoxType.PLAYER_TYPE,
    BoxType.PLAYER_COLOR_DISPOSITION,
    BoxType.PLAYER_SKILLS,
  ].includes(boxType);
```

File: packages/web/components/Quest/QuestDetails.tsx
Comment (dysbulic) [reply packages/web/components/Quest/QuestDetails.tsx]: In looking into the type information it doesn't look like it generates join associations.

That's such a pain to have to write & it makes for uglier code.

Oh well…

File: packages/web/graphql/getQuests.ts
Comment (dysbulic) [inline packages/web/graphql/getQuests.ts:128]:
```suggestion
  playerId: Scalars['uuid'],
```

**Review by dysbulic — APPROVED** [review]
I've not actually tested it b/c I don't know how to create or complete quests, but the code looks good to me.

File: hasura/metadata/tables.yaml
Comment (vidvidvid) [reply hasura/metadata/tables.yaml]: did it. but i didn't find any other cases with quests that would require this

File: packages/web/components/Quest/QuestFilter.tsx
Comment (vidvidvid) [reply packages/web/components/Quest/QuestFilter.tsx]: to not have to make 2 types 😅 
`type ValueType = { value: string; label: string };`

but i can make another one. makes more sense. :)

Comment (vidvidvid) [reply packages/web/components/Quest/QuestFilter.tsx]: actually the whole `MetaFilterSelectSearch` is set up by requiring strings as values. in the `MetaFilterSelectSearch the values are passed to `SelectSearch`, which also requires these values as strings. so instead of refactoring all of that, imo we can just leave it as it is?

Comment (dysbulic) [reply packages/web/components/Quest/QuestFilter.tsx]: Can you do `type ValueType = { value: string | number; label: string | number }` or `type ValueType<T = string> = { value: T; label: T }`? The second will have a string type by default, but you can use `ValueType<number>` to override it when you need to.

Comment (vidvidvid) [reply packages/web/components/Quest/QuestFilter.tsx]: renamed `ValueType` type to `FilterString`, just for clarity

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [reply packages/web/components/Quest/QuestFilter.tsx]: Fixing this would require digging into the type structure and arguments of `MetaFilterSelectSearch`. It is known that it takes an array of values, so the argument should be called `values` instead of `value`. The type passed into `onChange` needs some work as well.

It's not central to what you're doing at all & ought to be left for another day.

File: packages/web/components/Quest/QuestFilter.tsx
Comment (dysbulic) [reply packages/web/components/Quest/QuestFilter.tsx]: As I mentioned [elsewhere](//github.com/MetaFam/TheGame/pull/1085#discussion_r798335659), the brokenness of `MetaFilterSelectSearch` isn't something you have to deal with for this PR.

**Pull request merged**
